toolkit/pallets/address-associations/src/lib.rs

```diff
@@ -0,0 +1,107 @@
+//! Pallet storing associations from main chain addresses to parter chain address.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use alloc::string::ToString;
+	use frame_support::{fail, pallet_prelude::*};
+	use frame_system::pallet_prelude::OriginFor;
+	use sidechain_domain::{
+		MainchainAddress, MainchainKeyHash, MainchainPublicKey, MainchainSignature, UtxoId,
+	};
+	// use frame_system::pallet_prelude::*;
+	// use scale_info::prelude::fmt::Debug;
+	use sp_address_associations::AddressAssociationSignedMessage;
+	// use sp_std::vec::Vec;
+
+	pub const PALLET_VERSION: u32 = 1;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Partner chain address type
+		type PartnerChainAddress: Member + Parameter + MaxEncodedLen;
+
+		fn genesis_utxo() -> UtxoId;
+	}
+
+	#[pallet::storage]
+	pub type AddressAssociations<T: Config> = StorageMap<
+		Hasher = Blake2_128Concat,
+		Key = MainchainAddress,
+		Value = T::PartnerChainAddress,
+		QueryKind = OptionQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((2, DispatchClass::Normal))]
+		pub fn associate_address(
+			_origin: OriginFor<T>,
+			mc_addr: MainchainAddress,
+			pc_addr: T::PartnerChainAddress,
+			signature: MainchainSignature,
+			verification_key: MainchainPublicKey,
+		) -> DispatchResult {
+			let genesis_utxo = T::genesis_utxo();
+
+			let address_association_message = AddressAssociationSignedMessage {
+				mainchain_address: mc_addr.clone(),
+				partnerchain_address: pc_addr.clone(),
+				genesis_utxo,
+			};
+
+			let binary = address_association_message.encode();
+
+			let is_valid_signature = signature.verify(verification_key.clone(), &binary);
```

Comment:
try to avoid clone here, see other comment
```suggestion
			let is_valid_signature = signature.verify(&verification_key, &binary);
```